//trigger

trigger skedJobTrigger on sked__Job__c (after insert,after update) 
{
    try 
    {
        if (trigger.isInsert) skedJobTriggerHandler.onAfterInsert(trigger.new);
        if (trigger.isUpdate) skedJobTriggerHandler.onAfterUpdate(trigger.new, trigger.oldMap);
    }
     catch(DmlException e) 
    {
        System.debug('This is the following exception which has occurred: ' + e.getMessage());
    }
}




//class

public class skedJobTriggerHandler 
{
public static void onAfterInsert(List<sked__job__c> jobList) 
{
    sendJobCompletionSMS(jobList, null);
}    
public static void onAfterUpdate(List<sked__job__c> jobList, Map<id, sked__job__c> map_id_old) 
{
    sendJobCompletionSMS(jobList,map_id_old);
}     
public static void sendJobCompletionSMS(List<sked__job__c> jobList, Map<id, sked__job__c> map_id_old) 
{
    Set<Id> conIds = new Set<Id>();
    for (sked__Job__c job : jobList) 
    {
        if (map_id_old != null) 
        {
            if (job.sked__Job_Status__c != map_id_old.get(job.Id).sked__Job_Status__c && job.sked__Job_Status__c == 'Complete' && !String.isBlank(job.sked__Contact__c)) 
            {
                conIds.add(job.sked__Contact__c);
            }
        } 
        else if (job.sked__Job_Status__c == 'Complete' && !String.isBlank(job.sked__Contact__c)) 
        {
            conIds.add(job.sked__Contact__c);
        }
    }
    onCalloutResponse(conIds);
}
@future (callout=true) 
public static void onCalloutResponse(Set<Id> conIds) 
{        
    try 
    {
        List<Contact> conDetailList = [SELECT id, Phone, LastName FROM Contact WHERE Id IN : conIds LIMIT 5];
        if (conDetailList.size() > 0) 
        {
            String phone = conDetailList[0].Phone;
            String countryCode = 'IN';
            String message = 'handson exercise 2 is being processed';
            sked.ApiResult.Sms response = sked.SkeduloAPI.sms(message, countryCode, phone);
            if (response.isSuccess()) 
            {
                System.debug('Message has been sent--->'+response); 
            } 
            else 
            {
                System.debug('Error occured: ' + response.getError());
            } 
        }
    } 
    catch(Exception e) 
    {
        System.debug('This is the following exception which has occurred: ' + e.getMessage());
    }
    
}
}
